neuraxle/metaopt/tpe.py

```diff
@@ -0,0 +1,293 @@
+"""
+Tree parzen estimator
+====================================
+Code for tree parzen estimator auto ml.
+"""
+import numpy as np
+from typing import Optional
+from copy import deepcopy
+from neuraxle.hyperparams.space import HyperparameterSamples
+from neuraxle.hyperparams.distributions import DistributionMixture, Choice
+from .auto_ml import BaseHyperparameterOptimizer, RandomSearchHyperparameterOptimizer, TRIAL_STATUS
+from collections import Counter
+
+def _linear_forgetting_Weights(number_samples, number_recent_trial_at_full_weights):
+    """This part of code has been taken from Hyperopt (https://github.com/hyperopt) code."""
+    if number_samples == 0:
+        return np.asarray([])
+
+    if number_samples < number_recent_trial_at_full_weights:
+        return np.ones(number_samples)
+
+    weights_ramp = np.linspace(1.0 / number_samples, 1.0, number_samples - number_recent_trial_at_full_weights)
+    weights_flat = np.ones(number_recent_trial_at_full_weights)
+    weights = np.concatenate((weights_ramp, weights_flat), axis=0)
+
+    return weights
+
+
+class TreeParzenEstimatorHyperparameterOptimizer(BaseHyperparameterOptimizer):
+
+    def __init__(self, number_of_initial_random_step: int = 40, quantile_threshold: float = 30,
+                 number_possible_hyperparams_candidates=100,
+                 prior_weight: float = 0.,
+                 use_linear_forgetting_weights: bool = False,
+                 number_recent_trial_at_full_weights=25):
+        super().__init__()
+        self.initial_auto_ml_algo = RandomSearchHyperparameterOptimizer()
+        self.number_of_initial_random_step = number_of_initial_random_step
+        self.quantile_threshold = quantile_threshold
+        self.number_possible_hyperparams_candidates = number_possible_hyperparams_candidates
+        self.prior_weight = prior_weight
+        self.use_linear_forgetting_weights = use_linear_forgetting_weights
+        self.number_recent_trial_at_full_weights = number_recent_trial_at_full_weights
+
+    def _split_trials(self, success_trials):
```

Comment:
Because it is trials that has a success status. We see how to manage failed trial later on. `success_trials = auto_ml_container.trials.filter(TRIAL_STATUS.SUCCESS)`